### PR TITLE
Implement `Set` stdlib

### LINKED ISSUE #330: `Set` stdlib
Generally a `map` with `unit` value, but could be packed behind a nice lib

### PR DESCRIPTION
**Pull request by ghallak**
Fixes #330

### CODE REVIEW

**Review by radrow — CHANGES_REQUESTED** [review]
Looks fine, but please add docs in `docs/sophia_stdlib.md` (in the same manner it is done in all other modules). All public functions should be documented, even the obvious ones.

**Comment by ghallak** [comment]
@radrow Docs where added to `docs/sophia_stdlib.md` here 416999e19a541d421e2bc26c20ee6f44cc55eee6.

File: docs/sophia_stdlib.md [hunk 1]
```diff
@@ -2264,3 +2265,125 @@ accept arbitrary `frac`s from the surface you should report it as a
...
+
+Returns an empty set
+
+#### member
+
+```
+member(e : 'a, s : set('a)) : bool
+```
+
+Checks if the element is present in the Set
```
Comment (radrow) [inline docs/sophia_stdlib.md:2293]:
```suggestion
Checks if an element is present in the set
```
I think it works like that. Tbh, not sure – articles aren't part of my mother-tongue (sadly)

Comment (ghallak) [reply docs/sophia_stdlib.md]: Made some changes here 588684ba16cb8e6bc33b30f7fcd54f184ff02285

**Review by radrow — APPROVED** [review]

File: priv/stdlib/Set.aes [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    List.map(Pair.fst, Map.to_list(s.to_map))
+
+  function from_list(l : list('a)) : set('a) =
+    { to_map = Map.from_list(List.map((x) => (x, ()), l)) }
+  
+  function filter(p : 'a => bool, s : set('a)) : set('a) =
+    from_list(List.filter(p, to_list(s)))
+
+  function fold(f : ('a, 'b) => 'b, acc : 'b, s : set('a)) : 'b =
+    List.foldr(f, acc, to_list(l))
```
Comment (radrow) [inline priv/stdlib/Set.aes:33]:
```suggestion
    List.foldr(f, acc, to_list(s))
```

**Review by radrow — CHANGES_REQUESTED** [review]

File: priv/stdlib/Set.aes (hunk 2 above)
Comment (ghallak) [reply priv/stdlib/Set.aes]: Fixed here ed1c74cdcffb4dbe0bf532a286648043289a728f

**Review by radrow — APPROVED** [review]

File: docs/sophia_stdlib.md [hunk 3]
```diff
@@ -39,6 +39,7 @@ include "List.aes"
 - [Triple](#Triple)
 - [BLS12_381](#BLS12_381)
 - [Frac](#Frac)
+- [Set](#Set)
```
Comment (dincho) [inline docs/sophia_stdlib.md:42]: This actually points to Bits `set`

**Comment by dincho** [comment]
Don't you guys test the stdlib ?! 🤔

File: docs/sophia_stdlib.md (hunk 3 above)
Comment (radrow) [reply docs/sophia_stdlib.md:42]: Uh, it was always a mystery for me how do the labels in md work

**Comment by hanssv** [comment]
I hope there are tests, but you need FATE (or another VM :-) ) to run it, so those tests has to go in `aeternity` repo...

**Comment by radrow** [comment]
There are compilation tests, but no documentation tests (hello, machine learning team? are you around?). Afaik the runtime has been tested in REPL or aestudio

**Comment by ghallak** [comment]
I have written some tests in a dummy contract and made sure that they passed on aestudio. I will add the tests to aeternity repo and open a PR and link to it here.

File: docs/sophia_stdlib.md (hunk 3 above)
Comment (ghallak) [reply docs/sophia_stdlib.md]: This linking issue should be fixed in this commit 028b12ddc6c6a478207744057b6817b4abdfe0b2

**Review by radrow — DISMISSED** [review]
Nice trick with the ref

**Review by hanssv — DISMISSED** [review]
Looks good 👍

**Review by radrow — APPROVED** [review]
@hanssv I had to resolve a conflict, reclick pls

**Review by hanssv — APPROVED** [review]

**Pull request merged**